### PR TITLE
Ensure http response streams are paused

### LINKED ISSUE #437: Tracing destroys proxied payload
We've got an app that uses hapi with [h2o2](https://github.com/hapijs/h2o2) for proxying requests to a different service. When the trace agent is enabled, the responses seem to be empty. The response headers are included as expected, but there is no body.

I've made a reproducible test case here: https://github.com/rexxars/gtrace-proxy-issue

If I start the agent and disable *either* the `http` or the `hapi` instrumentation (or both), everything seems to work as expected.

### PR DESCRIPTION
**Pull request by matthewloring**
Responses returned by http#request are yielded in paused mode. By
attaching a 'data' listener to the request the http instrumentation was
switching the stream to flowing mode which could cause the request to
drain before the calling framework had a chance to attach their own
listeners.

Fixes #437

### CODE REVIEW

**Comment by ofrobots** [comment]
As discussed, it would be good to explore alternative ways to gather the request size rather than depending on internal state machine in `ReadableStream` – or to get reasonable confidence that this indeed is the best way to do this.

**Comment by matthewloring** [comment]
It looks like there are facilities for tracking the bytes read from a connection (https://nodejs.org/api/net.html#net_socket_bytesread) but connections can be reused and we don't observe all requests so we can't determine per-request size based on the bytes observed by the socket.

There is also the content-length header (https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) but servers are not required to supply this to my knowledge. 

As far as managing the stream state machine, I tried using `https://nodejs.org/api/stream.html#stream_readable_pause` but this doesn't restore the stream to its uninitialized state (https://nodejs.org/api/stream.html#stream_three_states). As a result the stream can buffer internally where it otherwise could not and with sufficient buffering may un-pause.

I can still investigate patches to `on`.

**Comment by matthewloring** [comment]
Here is a patch that will do the lazy initialization of our `data` listener. It's more wrapping/indirection but at least it doesn't dig into stream internals (even if they are documented officially). WDYT?

```
diff --git a/src/plugins/plugin-http.js b/src/plugins/plugin-http.js
index 355bcae..b8d5ea1 100644
--- a/src/plugins/plugin-http.js
+++ b/src/plugins/plugin-http.js
@@ -76,8 +76,22 @@ function patchRequest (http, api) {
       var req = request.call(this, options, function(res) {
         api.wrapEmitter(res);
         var numBytes = 0;
-        res.on('data', function (chunk) {
-          numBytes += chunk.length;
+        var listenerAttached = false;
+        // Responses returned by http#request are yielded in paused mode. By
+        // attaching a 'data' listener to the request the http instrumentation was
+        // switching the stream to flowing mode which could cause the request to
+        // drain before the calling framework had a chance to attach their own
+        // listeners. We now attach this listener lazily.
+        shimmer.wrap(res, 'on', function onWrap(on) {
+          return function on_trace() {
+            var eventName = arguments[0];
+            if (eventName === 'data' && !listenerAttached) {
+              on.call(this, 'data', function(chunk) {
+                numBytes += chunk.length;
+              });
+            }
+            on.apply(this, arguments);
+          };
         });
         res.on('end', function () {
           requestLifecycleSpan
```

**Comment by matthewloring** [comment]
As per offline discussion, we will go with the lazy initialization of our `on('data')` listener.

File: src/plugins/plugin-http.js [hunk 1]
```diff
@@ -76,8 +76,22 @@ function patchRequest (http, api) {
       var req = request.call(this, options, function(res) {
         api.wrapEmitter(res);
         var numBytes = 0;
-        res.on('data', function (chunk) {
-          numBytes += chunk.length;
+        var listenerAttached = false;
+        // Responses returned by http#request are yielded in paused mode. By
+        // attaching a 'data' listener to the request the http instrumentation was
+        // switching the stream to flowing mode which could cause the request to
```
Comment (ofrobots) [inline src/plugins/plugin-http.js]: nit: Phrase the comment in present tense. A future reader of the comment may not have to context about what used to happen. The comment should instead explain what is the intent of code as present.

File: src/plugins/plugin-http.js [hunk 2]
```diff
@@ -76,8 +76,22 @@ function patchRequest (http, api) {
       var req = request.call(this, options, function(res) {
         api.wrapEmitter(res);
         var numBytes = 0;
-        res.on('data', function (chunk) {
-          numBytes += chunk.length;
+        var listenerAttached = false;
```
Comment (ofrobots) [inline src/plugins/plugin-http.js]: Would this mechanism still work if the use calls `res.read()` rather than `res.on('data')` to consume the data?

Comment (matthewloring) [inline src/plugins/plugin-http.js]: Nope, added a comment to that effect.

File: src/plugins/plugin-http.js [hunk 3]
```diff
@@ -76,8 +76,22 @@ function patchRequest (http, api) {
...
+        // Responses returned by http#request are yielded in paused mode. By
+        // attaching a 'data' listener to the request the http instrumentation was
+        // switching the stream to flowing mode which could cause the request to
+        // drain before the calling framework had a chance to attach their own
+        // listeners. We now attach this listener lazily.
+        shimmer.wrap(res, 'on', function onWrap(on) {
+          return function on_trace() {
+            var eventName = arguments[0];
```
Comment (matthewloring) [inline src/plugins/plugin-http.js]: Done.

File: src/plugins/plugin-http.js (hunk 1 above)
Comment (matthewloring) [inline src/plugins/plugin-http.js]: done.

**Comment by ofrobots** [comment]
Lgtm

**Pull request merged**